bluemira/structural/element.py

```diff
@@ -0,0 +1,585 @@
+
+    __slots__ = (
+        "node_1",
+        "node_2",
+        "id_number",
+        "loads",
+        "stresses",
+        "max_stress",
+        "safety_factor",
+        "shapes",
+        "material",
+        "_properties",
+        "_material",
+        "_cross_section",
+        "_length",
+        "_weight",
+        "_k_matrix",
+        "_lambda_matrix",
+        "_k_matrix_glob",
+        "_s_functs",
+    )
+
+    def __init__(self, node_1, node_2, id_number, cross_section, material=None):
+        # Utility properties
+        self.node_1 = node_1
+        self.node_2 = node_2
+        self.id_number = id_number
+
+        # Public properties
+        self.loads = []
+
+        # Calculated properties
+        self.shapes = None
+        self.stresses = None
+        self.max_stress = None
+        self.safety_factor = None
+
+        # Private properties
+        self._material = material  # Record input material
+        self.material = None
+        self._properties = self._process_properties(cross_section, material)
+
+        self._cross_section = cross_section
+
+        # Cheap memo-isation of matrices (headache-free cached properties)
+        self._length = None
+        self._weight = None
+        self._k_matrix = None
+        self._lambda_matrix = None
+        self._k_matrix_glob = None
+
+        # Private construction utilities
+        self._s_functs = None
+
+    def _process_properties(self, cross_section, material):
+        """
+        Handles cross-sectional and material properties, including if a
+        composite material cross-section is specified.
+        """
+        properties = {}
+        if material is None:
+            # A composite cross-section was hopefully specified
+
+            properties["EA"] = cross_section.ea
+            properties["EIyy"] = cross_section.ei_yy
+            properties["EIzz"] = cross_section.ei_zz
+            properties["GJ"] = cross_section.gj
+            properties["nu"] = cross_section.nu
+            properties["ry"] = cross_section.ry
+            properties["rz"] = cross_section.rz
+            properties["A"] = cross_section.area
+            properties["rho"] = cross_section.rho  # area-weighted density
+
+            # Override material=None with a list of materials from the CS
+            self.material = cross_section.material
+        else:
+            # Single material weight cross-section properties
+            e_mat, g_mat = material.E, material.G
+            properties["EA"] = e_mat * cross_section.area
+            properties["EIyy"] = e_mat * cross_section.i_yy
+            properties["EIzz"] = e_mat * cross_section.i_zz
+            properties["GJ"] = g_mat * cross_section.j
+            properties["ry"] = cross_section.ry
+            properties["rz"] = cross_section.rz
+            properties["rho"] = material.rho
+            properties["A"] = cross_section.area
+            self.material = material
```

Comment:
I think my indentation is off
```suggestion
        # A composite cross-section was hopefully specified

        properties["nu"] = cross_section.nu
        properties["ry"] = cross_section.ry
        properties["rz"] = cross_section.rz
        properties["A"] = cross_section.area

        if material is None:
            properties["EA"] = cross_section.ea
            properties["EIyy"] = cross_section.ei_yy
            properties["EIzz"] = cross_section.ei_zz
            properties["GJ"] = cross_section.gj
            properties["rho"] = cross_section.rho  # area-weighted density

            # Override material=None with a list of materials from the CS
            self.material = cross_section.material
        else:
            # Single material weight cross-section properties
            e_mat, g_mat = material.E, material.G
            properties["EA"] = e_mat * cross_section.area
            properties["EIyy"] = e_mat * cross_section.i_yy
            properties["EIzz"] = e_mat * cross_section.i_zz
            properties["GJ"] = g_mat * cross_section.j
            properties["rho"] = material.rho
            self.material = material
```

Reply:
doesn't work like this; it's effectively to enable composite properties to be used (which is not the normal case)